activerecord/lib/active_record/enum.rb

```diff
@@ -168,7 +199,14 @@ def enum(definitions)
         end
 
         _enum_methods_module.module_eval do
-          pairs = values.respond_to?(:each_pair) ? values.each_pair : values.each_with_index
+          pairs = if values.respond_to?(:each_pair)
+            values.each_pair
+          elsif klass.table_name.present? && klass.column_for_attribute(name)&.type == :string
```

Comment:
This relates to a problem I had in my PR. I think this will break when defined on an abstract model.

Consider the following example present in the codebase:

```ruby
class Cat < ActiveRecord::Base
  self.abstract_class = true

  enum gender: [:female, :male]
end

class Lion < Cat
end
```

The column type for gender would be String, but is defined on Lion's table, while Cat's table isn't present because it's abstract. In this case, it would be defined as an integer and we don't won't that.

For this I had considered the following solutions, but I didn't implement them because I was waiting for @kamipo 's feedback at https://github.com/rails/rails/pull/33339#issuecomment-404068341 before dedicating more time in the PR:
1. Looking for `descendants` and then checking the column at the first descendent's table.
2. Specifying that it should be considered as a String by providing an option parameter.

I personally like more the first option, if you want I can also help you looking to this.

Reply:
@albertoalmagro That's true. I had the same problem first. The problem was caused by the absence of the `table_name` property in the model.

The thing is that the `column_for_attribute` method calls the `columns_hash` method which calles the `table_name` method. When the name of the table is not defined the SQL query for fetching available attributes becomes broken (with most all of available drivers) and the error will be raised. 

This situation is not likely possible with the production code, but in case it happens I've added a check for the `table_name` property presence.

Reply:
I do think it can happen in production. The problem is that if we define the enum on the abstract Model it will be treated as an integer enum instead of a string enum... I already imagine everyone blaming us 😄

Tomorrow or in the following days I can check this to see if I can find an easy solution 😉

Reply:
I think that the behaviour implemented in the code is a proper one. I'll try to explain why I do think so. Please tell me if I'm wrong anywhere :sweat_smile: 

Look. The test fails on the `Cat` model inclusion, not on the `Lion` model 
inclusion (take a look ant the `activerecord/test/cases/scoping/default_scoping_test.rb:9`):

```ruby
require "models/cat"
```

The `Lion` is inherited from the `Cat` model which is an abstract class without a real table behind it. And there's **absolutely no way to guess the type of the field** defined in `Cat`'s model **based on it's subclasses** because it can differ from class to class. It's not like STI logic there, just several models with different tables for each of them with a common superclass, marked as abstract. Take a look at the create statement in `test/schema/schema.rb`:

```ruby
  create_table :lions, force: true do |t|
    t.integer :gender
    t.boolean :is_vegetarian, default: false
  end
```

And the `Cat` model doesn't have it's own table, only the `lions` table exists for the `Lion` model (and pay attention, the `gender` attribute is defined as an `integer` there, but it doesn't matter in the context of the `Cat` model).
 
I think that the enum field should be treated as an integer if it isn't exactly declared as a string. **We shouldn't expand the string-logic on enums until we're 100% sure that the developer meant to use enum for a string field**. It is important for backwards-compatibility.

Reply:
There's another related issue with the current implementation: calling `column_for_attribute` here causes the schema to be queried while the class is being loaded, which is undesirable (https://github.com/rails/rails/commit/37661bfc810da7384ac6650d60be8669baa16813).

It seems to me that we should defer the decision of how to represent the enum value until `EnumType#serialize` is called:

https://github.com/rails/rails/blob/bd139a59405044c27bf367cde671e4bdbf7ccb57/activerecord/lib/active_record/enum.rb#L134-L136

That would solve the Cat/Lion problem described above, and also not load the schema prematurely.

Reply:
I think that your suggestion is reasonable. I'll fix my commit to satisfy it.